plugins/autoid/plugin.js

```diff
@@ -1,9 +1,95 @@
 'use strict';
 
 (function () {
-	CKEDITOR.plugins.add('autoid', {
-		init: function (editor) {
+  var editorHasFocus = false;
+  var commandIsActive = false;
+  var EVENT_NAMES = {
+    ALL_IDS_COMPLETE: 'allIdsComplete',
+    ID_ADDED: 'idAdded'
+  };
 
-		}
-	});
+  CKEDITOR.plugins.add('autoid', {
+    init: function (editor) {
+      var self = this;
+
+      editor.addFeature({
+        allowedContent: 'h1 h2 h3 h4 h5 h6 *[id]'
+      });
+
+      if (editor && !editor.config.autoid) {
+        editor.config.autoid = {};
+      }
+      self.settings = editor.config.autoid;
+      if (!self.settings) {
+        self.settings = {};
+      }
+
+      editor.addCommand('autoid', {
+        exec: function (editor) {
+          if (!commandIsActive) {
+            start();
+          } else {
+            stop();
+          }
+        },
+        editorFocus: true
+      });
+
+      editor.on("instanceReady", function () {
+        if (self.settings.autostart !== false) {
+          start()
+        }
+      });
+
+      editor.on('selectionChange', addIdIfNewHeading);
```

Comment:
I tried several different events.  `insertElement` doesn't fire when we need it to.  Generic `change` event fires for far more things than we need.  `selectionChange` fires whenever the selection start element changes.  It seems to capture all the necessary events without firing excessively otherwise.